src/psyclone/psyir/transformations/increase_rank_loop_arrays_trans.py

```diff
@@ -129,69 +130,101 @@ def validate(self, node, options=None):
-
-            if not array.is_automatic or not array.is_array:
+                try:
+                    array = node.scope.symbol_table.lookup(array)
+                except KeyError as err:
+                    raise TransformationError(
+                        f"{self.name} provided array '{array}' does not exist"
+                        f"in this scope."
+                    ) from err
+
+            if (not isinstance(array, Symbol) or not array.is_automatic
+                    or not array.is_array):
                 raise TransformationError(
-                    f"{self.name} provided 'arrays' must be a local array "
+                    f"{self.name} provided 'arrays' must be local array "
                     f"symbols, but '{array}' is not")
 
             if array.name.lower() in codeblock_symbols:
                 raise TransformationError(
                     f"{self.name} does not support arrays that are referenced "
                     f"inside a Codeblock, but '{array.name}' is inside one.")
+            # TODO: check that the bound expressions are valid as static
+            # expressions in the declarations.
 
-            # TODO: What should happen with references outside the loop?
-
-    def apply(self, node, options=None):
+    def apply(
+        self,
+        node: Loop,
+        arrays: Optional[list[Union[Symbol, str]]] = None,
+        **kwargs
+    ):
         '''Applies the transformation.
 
-        :param node: target PSyIR node.
-        :type node: :py:class:`psyclone.psyir.nodes.Loop`
-        :param options: a dictionary with options for transformations.
-        :type options: Optional[Dict[str, Any]]
+        :param node: target Loop node.
+        :param arrays: list of arrays that will have the rank increased.
 
         '''
-        self.validate(node, options)
-
-        array_list = options.get("arrays", [])
+        self.validate(node, arrays, **kwargs)
 
-        for array in array_list:
+        for array in arrays:
             if isinstance(array, str):
                 array = node.scope.symbol_table.lookup(array, otherwise=None)
-            if array is None:
-                continue
 
-            # TODO: check that the bound expressions are valid as static
-            # expressions in the declarations.
             array.shape.append(
                 ArrayType.ArrayBounds(node.start_expr, node.stop_expr)
             )
 
-            for ref in node.walk(ArrayReference):
+            for ref in array.find_symbol_table(node).node.walk(ArrayReference):
                 if ref.symbol is array:
-                    ref.addchild(Reference(node.variable))
+                    if ref.is_descendent_of(node):
+                        # Inside the target loop index the reference to the
+                        # loop variable
+                        ref.addchild(Reference(node.variable))
+                    else:
+                        # Outside the target loop index the whole range
```

Comment:
What about if the scoping region contains two explicit loops that both use the same array (before its rank is increased)? I think this will end up transforming the accesses inside a second loop to use full slices. We should probably catch and reject this case in `validate`.

Reply:
> I think this will end up transforming the accesses inside a second loop to use full slices.
Yes it will. But I think this is the right behaviour, see last example with the initialisation test. We essentially trade more parallelism (independent elements), by duplicating this elements outside the loop.

Reply:
Ok, I partially came around on this since our discussion and now I am much more restrictive in that the validate only allows simple initialisation expressions "ztmp(..) = ..." outside the loop. As evaluating the correctness of other cases was very complex. I think with this I don't need the warning in the docstring, as now I am more confident that the semantics do not change. And the previous condition already blocks applying it two times, as it will find the array usage in the other loop unless they are simple write-only loops, in which case is also semantically equivalent. Let me know what you think.